packages/node/src/login/oidc/redirectHandler/AuthCodeRedirectHandler.ts

```diff
@@ -74,8 +199,89 @@ export class AuthCodeRedirectHandler implements IRedirectHandler {
       );
     }
 
-    throw new NotImplementedError(
-      "AuthCodeRedirectHandler not implemented for Node"
+    const url = new URL(redirectUrl);
+    // The type assertion is ok, because we checked in canHandle for the presence of a state
+    const oauthState = url.searchParams.get("state") as string;
+
+    const {
+      sessionId,
+      issuerConfig,
+      codeVerifier,
+      redirectUri,
+      dpop,
+    } = await retrieveContextFromStorage(
+      oauthState,
+      this.storageUtility,
+      this.issuerConfigFetcher
+    );
+
+    const issuer = new Issuer(configToIssuerMetadata(issuerConfig));
+
+    // This should also retrieve the client from storage
+    const clientInfo: IClient = await this.clientRegistrar.getClient(
+      { sessionId },
+      issuerConfig
+    );
+
+    const client = new issuer.Client({
+      client_id: clientInfo.clientId,
+      client_secret: clientInfo.clientSecret,
+    });
+
+    const params = client.callbackParams(redirectUrl);
+    let dpopKey: JWK.ECKey;
+    let tokens: TokenSet;
+    let authFetch: typeof fetch;
+
+    if (dpop) {
+      dpopKey = await JWK.generate("EC", "P-256");
+      tokens = await client.callback(
+        redirectUri,
+        params,
+        { code_verifier: codeVerifier },
+        { DPoP: dpopKey.toJWK(true) }
+      );
+    } else {
+      tokens = await client.callback(redirectUri, params, {
+        code_verifier: codeVerifier,
+      });
+    }
+    if (tokens.access_token === undefined || tokens.id_token === undefined) {
+      // The error message is left minimal on purpose not to leak the tokens.
+      throw new Error("The IdP did not return the expected tokens.");
+    }
+    if (dpop) {
+      authFetch = await buildDpopFetch(
+        tokens.access_token,
+        tokens.refresh_token,
+        // TODO: buildDpopFetch isn't implemented yet
+        // TS thinks dpopKey isn't initialized, when it is.
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        dpopKey as any
+      );
+    } else {
+      authFetch = buildBearerFetch(tokens.access_token, tokens.refresh_token);
+    }
+
+    const webid = await deriveWebidFromTokenPayload(tokens.claims());
```

Comment:
```suggestion
    const webid = await deriveWebIdFromTokenPayload(tokens.claims());
```

Reply:
I used 'derive' because it is used in the [WebID-OIDC spec](https://github.com/solid/webid-oidc-spec#deriving-webid-uri-from-id-token), but I'm fine with anything else. I used "payload" specifically, because we don't parse the entire ID Token, which also has headers and a signature, so payload here comes from JWT. I'd prefer to keep it in, because we on purpose don't parse the whole ID Token (that's what `.claims()` does, and it also verifies the signature).

I changed `tokens` to `tokenSet`, as well as a comment about what `.claims()` returns as you suggested.

Reply:
Resolved in 687516df762ab9f8dbbaae9605bf250ac66de6dd